daemon/emer-aggregate-timer-impl.c

```diff
@@ -0,0 +1,165 @@
+
+/* Copyright 2021 Endless OS Foundation, LLC */
+
+/*
+ * This file is part of eos-event-recorder-daemon.
+ *
+ * eos-event-recorder-daemon is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * eos-event-recorder-daemon is distributed in the hope that it will be
+ * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with eos-event-recorder-daemon.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include "emer-aggregate-timer-impl.h"
+#include "emer-aggregate-tally.h"
+
+struct _EmerAggregateTimerImpl
+{
+  GObject parent_instance;
+
+  EmerAggregateTally *tally; /* unowned */
+  gint64 start_monotonic_us;
+  gint64 end_monotonic_us;
+
+  guint32 unix_user_id;
+  GVariant *event_id; /* owned */
+  GVariant *aggregate_key; /* owned */
+  GVariant *payload; /* owned */
+  GVariant *cache_key; /* owned */
+};
+
+G_DEFINE_TYPE (EmerAggregateTimerImpl, emer_aggregate_timer_impl, G_TYPE_OBJECT)
+
+static void
+emer_aggregate_timer_impl_finalize (GObject *object)
+{
+  EmerAggregateTimerImpl *self = (EmerAggregateTimerImpl *)object;
+
+  g_clear_pointer (&self->event_id, g_variant_unref);
+  g_clear_pointer (&self->aggregate_key, g_variant_unref);
+  g_clear_pointer (&self->payload, g_variant_unref);
+  g_clear_pointer (&self->cache_key, g_variant_unref);
+
+  G_OBJECT_CLASS (emer_aggregate_timer_impl_parent_class)->finalize (object);
+}
+
+static void
+emer_aggregate_timer_impl_class_init (EmerAggregateTimerImplClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = emer_aggregate_timer_impl_finalize;
+}
+
+static void
+emer_aggregate_timer_impl_init (EmerAggregateTimerImpl *self)
+{
+}
+
+EmerAggregateTimerImpl *
+emer_aggregate_timer_impl_new (EmerAggregateTally *tally,
+                               guint32             unix_user_id,
+                               GVariant           *event_id,
+                               GVariant           *aggregate_key,
+                               GVariant           *payload,
+                               gint64              monotonic_time_us)
+{
+  EmerAggregateTimerImpl *self;
+  g_autoptr(GVariant) cache_key = NULL;
+
+  self = g_object_new (EMER_TYPE_AGGREGATE_TIMER_IMPL, NULL);
+  self->tally = tally;
+  self->unix_user_id = unix_user_id;
+  self->event_id = g_variant_ref (event_id);
+  self->aggregate_key = g_variant_ref (aggregate_key);
+  self->payload = payload ? g_variant_ref (payload) : NULL;
+  self->start_monotonic_us = monotonic_time_us;
+  self->cache_key = g_variant_new ("(u@ayvmv)",
```

Comment:
This variant is used to hash the aggregate timer, but it contains `mv` - is that a problem?

Reply:
Yes, although the problem is actually that the variant is a container (tuple) at all — `g_variant_hash()` is defined only for basic variant types (i.e. not containers) and will emit a runtime warning when called on this.

You’ll need to build your own hash function, or choose a basic variant from within the cache key which can hash and represent the whole thing.

It’s fine to pass this type to `g_variant_equal()` though. In general, `maybe` types are fully supported by GVariant except for passing over D-Bus.

Reply:
Do we have any cases where the payload is neither None nor the same as the aggregate key? I don't think so – [Daily|Monthly]AppUsage uses the same string, and time-in-session has None payload. So perhaps we can merge them in the D-Bus API as just the aggregate key + boolean for whether to send it as the payload to the server, and avoid the need to have a maybe here?

Reply:
I might be misunderstanding what you’re suggesting, so to reiterate:
 - If you call `g_variant_hash()` on a variant of type `(u@ayvmv)` it will fail because the variant is a tuple, not because it contains a `maybe`
 - Calling `g_variant_hash()` on `(u@ayvsb)` or `(u@ayvv)` or `(*)` will fail similarly

Reply:
I had misunderstood. Thanks!

Reply:
I have worked around this by printing the cache key variant into a string, and comparing the strings. I'm assuming that two variants built equally should print the same string.

Reply:
> I'm assuming that two variants built equally should print the same string.

Yes, if you are using `g_variant_print()`. That’s actually exactly how `g_variant_equal()` works!